Auth server panics on gRPC request

We are occasionally encountering this panic on Teleport Cloud tenants:
```

2023-11-28 07:21:36.465	
{"caller":"auth/grpcserver.go:2024","component":"auth","level":"debug","message":"Client version \"13.1.0\" is less than 14.0.0, converting role to v6","timestamp":"2023-11-28T12:21:36Z"}
2023-11-28 07:21:36.466	
{"caller":"auth/grpcserver.go:2038","component":"auth","level":"debug","message":"Downgrading role \"[REDACTED]\" before returning it to the client: Client version \"13.1.0\" does not support Role v7. Role \"[REDACT]\" will be downgraded by adding more stringent restriction rules for Kubernetes clusters which will affect its behavior before returning to the client. In order to guarantee the correct behavior, all clients must be updated to version \"14.0.0-aa\" or higher.","timestamp":"2023-11-28T12:21:36Z"}

2023-11-28 07:21:36.469	
panic: runtime error: slice bounds out of range [-261:]
2023-11-28 07:21:36.469	

2023-11-28 07:21:36.469	
goroutine 32729 [running]:
2023-11-28 07:21:36.469	
github.com/gravitational/teleport/api/types.(*Metadata).MarshalToSizedBuffer(0xc0083d9030, {0xc0018d4f00, 0x65, 0x277})
2023-11-28 07:21:36.469	
	github.com/gravitational/teleport/api@v0.0.0/types/types.pb.go:19776 +0xa35
2023-11-28 07:21:36.469	
github.com/gravitational/teleport/api/types.(*RoleV6).MarshalToSizedBuffer(0xc0083d9000, {0xc0018d4f00, 0x277, 0x277})

2023-11-28 07:21:36.469	
	github.com/gravitational/teleport/api@v0.0.0/types/types.pb.go:26921 +0x185
2023-11-28 07:21:36.469	
github.com/gravitational/teleport/api/types.(*RoleV6).Marshal(0x8a90780?)
2023-11-28 07:21:36.469	
	github.com/gravitational/teleport/api@v0.0.0/types/types.pb.go:26889 +0x4d
2023-11-28 07:21:36.469	
google.golang.org/protobuf/internal/impl.legacyMarshal({{}, {0xa65db08, 0xc0050c3330}, {0x0, 0x0, 0x0}, 0x0})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/internal/impl/legacy_message.go:402 +0x9e
2023-11-28 07:21:36.469	
google.golang.org/protobuf/proto.MarshalOptions.size({{}, 0xe?, 0x2c?, 0x83?}, {0xa65db08, 0xc0050c3330})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/proto/size.go:43 +0xa6
2023-11-28 07:21:36.469	
google.golang.org/protobuf/proto.MarshalOptions.Size({{}, 0x96?, 0x1?, 0x0?}, {0xa597280?, 0xc0050c3330?})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/proto/size.go:26 +0x4c
2023-11-28 07:21:36.469	
google.golang.org/protobuf/proto.Size(...)
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/proto/size.go:16
2023-11-28 07:21:36.469	
google.golang.org/protobuf/internal/impl.sizeMessage({0xa597280?, 0xc0050c3330?}, 0x1, {0xb0?})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/internal/impl/codec_field.go:266 +0x2e
2023-11-28 07:21:36.469	
google.golang.org/protobuf/internal/impl.makeMessageFieldCoder.func1({0x1?}, 0xc000b93950, {0x0?})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/internal/impl/codec_field.go:209 +0x4d
2023-11-28 07:21:36.469	
google.golang.org/protobuf/internal/impl.(*MessageInfo).initOneofFieldCoders.func3({0xa54ebe8?}, 0xb0a058?, {0x1?})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/internal/impl/codec_field.go:89 +0x38
2023-11-28 07:21:36.469	
google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow(0xc000fda580, {0x40?}, {0x40?})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/internal/impl/encode.go:79 +0xd3
2023-11-28 07:21:36.469	
google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer(0x8?, {0xc0020a0000?}, {0x80?})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/internal/impl/encode.go:56 +0x6c
2023-11-28 07:21:36.469	
google.golang.org/protobuf/internal/impl.(*MessageInfo).size(0x427b45?, {{}, {0xa65ced8?, 0xc0024c3300?}, 0x79?})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/internal/impl/encode.go:40 +0x52
2023-11-28 07:21:36.469	
google.golang.org/protobuf/proto.MarshalOptions.size({{}, 0x98?, 0x3e?, 0x65?}, {0xa65ced8, 0xc0024c3300})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/proto/size.go:35 +0xc6
2023-11-28 07:21:36.469	
google.golang.org/protobuf/proto.MarshalOptions.Size({{}, 0xb0?, 0x15?, 0x62?}, {0xa570e40?, 0xc0024c3300?})
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/proto/size.go:26 +0x4c
2023-11-28 07:21:36.469	
google.golang.org/protobuf/proto.Size(...)
2023-11-28 07:21:36.469	
	google.golang.org/protobuf@v1.31.0/proto/size.go:16
2023-11-28 07:21:36.469	
github.com/gravitational/teleport/lib/auth.(*GRPCServer).WatchEvents(0x0?, 0xc00656e980, {0xa6355a0, 0xc001336730})
2023-11-28 07:21:36.469	
	github.com/gravitational/teleport/lib/auth/grpcserver.go:412 +0x3a6
2023-11-28 07:21:36.469	
github.com/gravitational/teleport/api/client/proto._AuthService_WatchEvents_Handler({0x8ae0860?, 0xc001e7e360}, {0xa62ecb0, 0xc00347a940})
2023-11-28 07:21:36.469	
	github.com/gravitational/teleport/api@v0.0.0/client/proto/authservice.pb.go:19947 +0xd3
2023-11-28 07:21:36.469	
github.com/gravitational/teleport/lib/auth.(*Middleware).withAuthenticatedUserStreamInterceptor(0x0?, {0x8ae0860, 0xc001e7e360}, {0xa62fcb0, 0xc001336690}, 0xc0043333b0?, 0x99e9bc0)
2023-11-28 07:21:36.469	
	github.com/gravitational/teleport/lib/auth/middleware.go:468 +0xdf
2023-11-28 07:21:36.469	
google.golang.org/grpc.getChainStreamHandler.func1({0x8ae0860, 0xc001e7e360}, {0xa62fcb0, 0xc001336690})
2023-11-28 07:21:36.469	
	google.golang.org/grpc@v1.58.3/server.go:1520 +0xb2
2023-11-28 07:21:36.469	
github.com/gravitational/teleport/lib/limiter.(*Limiter).StreamServerInterceptor(0xc001cdc500, {0x8ae0860, 0xc001e7e360}, {0xa62fcb0, 0xc001336690?}, 0x10?, 0xc00656e940)
2023-11-28 07:21:36.469	
	github.com/gravitational/teleport/lib/limiter/limiter.go:176 +0x1c6
2023-11-28 07:21:36.469	
google.golang.org/grpc.getChainStreamHandler.func1({0x8ae0860, 0xc001e7e360}, {0xa62fcb0, 0xc001336690})
2023-11-28 07:21:36.469	
	google.golang.org/grpc@v1.58.3/server.go:1520 +0xb2
2023-11-28 07:21:36.469	
github.com/gravitational/teleport/api/utils/grpc/interceptors.GRPCServerStreamErrorInterceptor({0x8ae0860, 0xc001e7e360}, {0xa62e7e8?, 0xc001a83020}, 0xc003e64f90?, 0xc00656e900)
2023-11-28 07:21:36.469	
	github.com/gravitational/teleport/api@v0.0.0/utils/grpc/interceptors/errors.go:76 +0x83
2023-11-28 07:21:36.469	
google.golang.org/grpc.getChainStreamHandler.func1({0x8ae0860, 0xc001e7e360}, {0xa62e7e8, 0xc001a83020})
2023-11-28 07:21:36.469	
	google.golang.org/grpc@v1.58.3/server.go:1520 +0xb2
2023-11-28 07:21:36.469	
github.com/gravitational/teleport/lib/auth.(*Middleware).StreamInterceptors.StreamServerInterceptor.StreamServerInterceptor.func1({0x8ae0860, 0xc001e7e360}, {0xa62fa70, 0xc001a82ea0}, 0xc003e64f90, 0xc00656e8c0)
2023-11-28 07:21:36.469	
	github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20220308023801-e4a6915ea237/interceptors/server.go:35 +0x2e8
2023-11-28 07:21:36.469	
google.golang.org/grpc.getChainStreamHandler.func1({0x8ae0860, 0xc001e7e360}, {0xa62fa70, 0xc001a82ea0})
2023-11-28 07:21:36.469	
	google.golang.org/grpc@v1.58.3/server.go:1520 +0xb2
2023-11-28 07:21:36.469	
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.StreamServerInterceptor.func1({0x8ae0860, 0xc001e7e360}, {0xa62f4d8, 0xc00334ed20}, 0xc003e64f90, 0xc00656e7c0)
2023-11-28 07:21:36.469	
	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.42.0/interceptor.go:437 +0x431
2023-11-28 07:21:36.469	
google.golang.org/grpc.NewServer.chainStreamServerInterceptors.chainStreamInterceptors.func2({0x8ae0860, 0xc001e7e360}, {0xa62f4d8, 0xc00334ed20}, 0x7469200?, 0xc001336170?)
2023-11-28 07:21:36.469	
	google.golang.org/grpc@v1.58.3/server.go:1511 +0x85
2023-11-28 07:21:36.469	
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0022150e0, {0xa63f840, 0xc0001836c0}, 0xc0031acea0, 0xc001aca1b0, 0xf3894c0, 0x0)
2023-11-28 07:21:36.469	
	google.golang.org/grpc@v1.58.3/server.go:1675 +0x12c7
2023-11-28 07:21:36.469	
google.golang.org/grpc.(*Server).handleStream(0xc0022150e0, {0xa63f840, 0xc0001836c0}, 0xc0031acea0, 0x0)
2023-11-28 07:21:36.469	
	google.golang.org/grpc@v1.58.3/server.go:1755 +0x9a6
2023-11-28 07:21:36.469	
google.golang.org/grpc.(*Server).serveStreams.func1.1()
2023-11-28 07:21:36.469	
	google.golang.org/grpc@v1.58.3/server.go:986 +0xbb
2023-11-28 07:21:36.469	
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 32402
2023-11-28 07:21:36.469	
	google.golang.org/grpc@v1.58.3/server.go:997 +0x145
```

Found a vaguely similar error here, noting that opentelemetry is in the callstack: https://github.com/open-telemetry/opentelemetry-collector/issues/6420

--- a/lib/auth/grpcserver.go
+++ b/lib/auth/grpcserver.go
@@ -2091,6 +2091,11 @@ func maybeDowngradeRoleLabelExpressions(ctx context.Context, role *types.RoleV6,
 	if !clientVersion.LessThan(minSupportedLabelExpressionVersion) {
 		return role, nil
 	}
+	// Make a shallow copy of the role so that we don't mutate the original.
+	// This is necessary because the role is stored in the backend and it's shared
+	// between multiple clients sessions.
+	// If we mutate the original role, it will be mutated for all clients
+	// which can cause panics since it causes a race condition.
 	role = apiutils.CloneProtoMsg(role)
 	hasLabelExpression := false
 	for _, kind := range types.LabelMatcherKinds {
@@ -2166,6 +2171,11 @@ func downgradeRoleToV6(r *types.RoleV6) (*types.RoleV6, bool, error) {
 		return r, false, nil
 	case types.V7:
 		var restricted bool
+		// Make a shallow copy of the role so that we don't mutate the original.
+		// This is necessary because the role is stored in the backend and it's shared
+		// between multiple clients sessions.
+		// If we mutate the original role, it will be mutated for all clients
+		// which can cause panics since it causes a race condition.
 		downgraded := apiutils.CloneProtoMsg(r)
 		downgraded.Version = types.V6
 

--- a/lib/auth/grpcserver.go
+++ b/lib/auth/grpcserver.go
@@ -2092,8 +2092,8 @@ func maybeDowngradeRoleLabelExpressions(ctx context.Context, role *types.RoleV6,
 		return role, nil
 	}
 	// Make a shallow copy of the role so that we don't mutate the original.
-	// This is necessary because the role is stored in the backend and it's shared
-	// between multiple clients sessions.
+	// This is necessary because the role is shared
+	// between multiple clients sessions when notifying about changes in watchers.
 	// If we mutate the original role, it will be mutated for all clients
 	// which can cause panics since it causes a race condition.
 	role = apiutils.CloneProtoMsg(role)
@@ -2172,8 +2172,8 @@ func downgradeRoleToV6(r *types.RoleV6) (*types.RoleV6, bool, error) {
 	case types.V7:
 		var restricted bool
 		// Make a shallow copy of the role so that we don't mutate the original.
-		// This is necessary because the role is stored in the backend and it's shared
-		// between multiple clients sessions.
+		// This is necessary because the role is shared
+		// between multiple clients sessions when notifying about changes in watchers.
 		// If we mutate the original role, it will be mutated for all clients
 		// which can cause panics since it causes a race condition.
 		downgraded := apiutils.CloneProtoMsg(r)

--- a/lib/auth/grpcserver.go
+++ b/lib/auth/grpcserver.go
@@ -61,6 +61,7 @@ import (
 	apievents "github.com/gravitational/teleport/api/types/events"
 	"github.com/gravitational/teleport/api/types/installers"
 	"github.com/gravitational/teleport/api/types/wrappers"
+	apiutils "github.com/gravitational/teleport/api/utils"
 	"github.com/gravitational/teleport/lib/auth/assist/assistv1"
 	"github.com/gravitational/teleport/lib/auth/discoveryconfig/discoveryconfigv1"
 	integrationService "github.com/gravitational/teleport/lib/auth/integration/integrationv1"
@@ -2090,6 +2091,7 @@ func maybeDowngradeRoleLabelExpressions(ctx context.Context, role *types.RoleV6,
 	if !clientVersion.LessThan(minSupportedLabelExpressionVersion) {
 		return role, nil
 	}
+	role = apiutils.CloneProtoMsg(role)
 	hasLabelExpression := false
 	for _, kind := range types.LabelMatcherKinds {
 		allowLabelMatchers, err := role.GetLabelMatchers(types.Allow, kind)
@@ -2163,11 +2165,8 @@ func downgradeRoleToV6(r *types.RoleV6) (*types.RoleV6, bool, error) {
 	case types.V3, types.V4, types.V5, types.V6:
 		return r, false, nil
 	case types.V7:
-		var (
-			downgraded types.RoleV6
-			restricted bool
-		)
-		downgraded = *r
+		var restricted bool
+		downgraded := apiutils.CloneProtoMsg(r)
 		downgraded.Version = types.V6
 
 		if len(downgraded.GetKubeResources(types.Deny)) > 0 {
@@ -2225,7 +2224,7 @@ func downgradeRoleToV6(r *types.RoleV6) (*types.RoleV6, bool, error) {
 			restricted = true
 		}
 
-		return &downgraded, restricted, nil
+		return downgraded, restricted, nil
 	default:
 		return nil, false, trace.BadParameter("unrecognized role version %T", r.Version)
 	}

--- a/lib/auth/grpcserver_test.go
+++ b/lib/auth/grpcserver_test.go
@@ -1808,10 +1808,9 @@ func TestIsMFARequired(t *testing.T) {
 
 					// If auth pref or role require session MFA, and MFA is not already
 					// verified according to private key policy, expect MFA required.
-					wantRequired :=
-						(role.GetOptions().RequireMFAType.IsSessionMFARequired() || authPref.GetRequireMFAType().IsSessionMFARequired()) &&
-							!role.GetPrivateKeyPolicy().MFAVerified() &&
-							!authPref.GetPrivateKeyPolicy().MFAVerified()
+					wantRequired := (role.GetOptions().RequireMFAType.IsSessionMFARequired() || authPref.GetRequireMFAType().IsSessionMFARequired()) &&
+						!role.GetPrivateKeyPolicy().MFAVerified() &&
+						!authPref.GetPrivateKeyPolicy().MFAVerified()
 					var wantMFARequired proto.MFARequired
 					if wantRequired {
 						wantMFARequired = proto.MFARequired_MFA_REQUIRED_YES
@@ -3932,6 +3931,9 @@ func TestRoleVersions(t *testing.T) {
 
 	newRole := func(version string, spec types.RoleSpecV6) types.Role {
 		role, err := types.NewRoleWithVersion("test_rule", version, spec)
+		meta := role.GetMetadata()
+		meta.Labels = map[string]string{"env": "staging"}
+		role.SetMetadata(meta)
 		require.NoError(t, err)
 		return role
 	}
@@ -4074,6 +4076,10 @@ func TestRoleVersions(t *testing.T) {
 						// and ignore it in the role diff.
 						if tc.expectDowngraded {
 							require.NotEmpty(t, gotRole.GetMetadata().Labels[types.TeleportDowngradedLabel])
+							require.NotSame(t, role, gotRole)
+							// The labels map is a pointer, so make sure it's was properly
+							// cloned.
+							require.NotSame(t, role.GetMetadata().Labels, gotRole.GetMetadata().Labels)
 						}
 					}
 					checkErr := func(err error) {

--- a/lib/auth/grpcserver_test.go
+++ b/lib/auth/grpcserver_test.go
@@ -23,6 +23,7 @@ import (
 	"encoding/pem"
 	"fmt"
 	"io"
+	"maps"
 	"net"
 	"net/http"
 	"net/http/httptest"
@@ -30,6 +31,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/coreos/go-semver/semver"
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
 	"github.com/google/uuid"
@@ -3929,10 +3931,11 @@ func TestRoleVersions(t *testing.T) {
 
 	wildcardLabels := types.Labels{types.Wildcard: {types.Wildcard}}
 
+	originalLabels := map[string]string{"env": "staging"}
 	newRole := func(version string, spec types.RoleSpecV6) types.Role {
 		role, err := types.NewRoleWithVersion("test_rule", version, spec)
 		meta := role.GetMetadata()
-		meta.Labels = map[string]string{"env": "staging"}
+		meta.Labels = maps.Clone(originalLabels)
 		role.SetMetadata(meta)
 		require.NoError(t, err)
 		return role
@@ -4076,10 +4079,6 @@ func TestRoleVersions(t *testing.T) {
 						// and ignore it in the role diff.
 						if tc.expectDowngraded {
 							require.NotEmpty(t, gotRole.GetMetadata().Labels[types.TeleportDowngradedLabel])
-							require.NotSame(t, role, gotRole)
-							// The labels map is a pointer, so make sure it's was properly
-							// cloned.
-							require.NotSame(t, role.GetMetadata().Labels, gotRole.GetMetadata().Labels)
 						}
 					}
 					checkErr := func(err error) {
@@ -4155,6 +4154,15 @@ func TestRoleVersions(t *testing.T) {
 						} else {
 							require.NoError(t, err)
 						}
+					}
+
+					// Call maybeDowngrade directly to make sure the original
+					// role isn't modified
+					sv, err := semver.NewVersion(clientVersion)
+					if err == nil {
+						_, err := maybeDowngradeRoleToV6(ctx, role.(*types.RoleV6), sv)
+						require.NoError(t, err)
+						require.Equal(t, originalLabels, role.GetMetadata().Labels)
 					}
 				})
 			}

